File: bend2/effs/udp_bind.c
// UDP
// ===

Term udp_bind_run(Env e, Term* f, IoWork* w) {
  int fd = socket(AF_INET, SOCK_DGRAM, 0);
  if (fd < 0) {
    return io_fail(e, (uint32_t)errno, NULL);
  }
  struct sockaddr_in at;
  if (io_sys_addr("0.0.0.0", (uint32_t)f[0], &at) < 0) {
    close(fd);
    return io_fail(e, EINVAL, NULL);
  }
  if (bind(fd, (struct sockaddr*)&at, sizeof(at)) < 0
    || fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0) {
    uint32_t code = (uint32_t)errno;
    close(fd);
    return io_fail(e, code, NULL);
  }
  return io_done(e, io_hand(fd));
}

static void __attribute__((constructor)) udp_bind_use(void) {
  io_eff(CID(UDP.bind), udp_bind_run, 0);
}

File: bend2/effs/udp_bind.js
// UDP
// ===

function udp_bind(port) {
  const sys = io_sys();
  const fd = sys.socket(2, 2, 0);
  if (fd < 0) {
    return io_fail(sys.errno());
  }
  const at = io_addr("0.0.0.0", Number(port));
  if (at === null) {
    sys.close(fd);
    return io_fail(22);
  }
  if (sys.bind(fd, sys.ptr(at), 16) < 0) {
    const code = sys.errno();
    sys.close(fd);
    return io_fail(code);
  }
  const h = io_mint("Socket", "udp", fd);
  if (h === null) {
    sys.close(fd);
    return io_fail(24);
  }
  return io_done(h);
}
